### PR TITLE
Ticket28755 v2: K=V parsing with fuzzer

### PR DESCRIPTION
**Pull request by nmathewson**

### CODE REVIEW

File: src/lib/encoding/kvline.c [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+#include "lib/string/printf.h"
+#include "lib/string/util_string.h"
+#include "lib/log/escape.h"
+#include "lib/log/util_bug.h"
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <string.h>
+
+#include <stdio.h>//XXXX
```
Comment (dgoulet-tor) [inline src/lib/encoding/kvline.c]: rm that :)

File: src/lib/encoding/kvline.c [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    if (esc) {
+      tmp = esc_for_log(line->value);
+      v = tmp;
+    } else {
+      v = line->value;
+    }
+
+    //printf("  <%s!%s!%s>::\n", k,eq,v);//XXXX
```
Comment (dgoulet-tor) [inline src/lib/encoding/kvline.c]: Leftovers

File: src/lib/encoding/kvline.c [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    smartlist_add_asprintf(elements, "%s%s%s", k, eq, v);
+    tor_free(tmp);
+  }
+
+  char *result = smartlist_join_strings(elements, " ", 0, NULL);
+
+  //printf("::%s::\n", result);//XXXX
```
Comment (dgoulet-tor) [inline src/lib/encoding/kvline.c]: Leftovers

File: src/lib/encoding/kvline.c [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+  return result;
+}
+
+/**
+ * Decode a <b>line</b> containing a series of space-separated 'Key=Value'
+ * pairs, using the provided <b>flags</b> to decode it.  Return a newly
+ * allocated list of pairs on success, or NULL on failure.
+ *
+ * If KV_QUOTED is set in <b>flags</b>, then quoted values are
+ * allowed. Otherwise, such values are not allowed.
```
Comment (dgoulet-tor) [inline src/lib/encoding/kvline.c]: The single quote is not parsed so `KV_QUOTED` implies double quotes. Is the term "Quoted" usually implies only double quotes?

File: src/test/test_config.c [hunk 5]
```diff
@@ -5813,6 +5814,71 @@ test_config_extended_fmt(void *arg)
...
+  enc = kvline_encode(lines, 0);
+  tt_assert(!enc);
+  enc = kvline_encode(lines, KV_QUOTED|KV_OMIT_KEYS);
+  tt_str_op(enc, OP_EQ, "AB=C CDE=\"F G\" GHI");
+  tor_free(enc);
+  config_free_lines(lines);
+
+  lines = kvline_parse("A\"B=C CDE=\"F\" \"GHI\" ", KV_QUOTED|KV_OMIT_KEYS);
+  tt_assert(! lines);
+
```
Comment (dgoulet-tor) [inline src/test/test_config.c]: So I've tried this and the first one works but the second one fails where I indicated. I'm guessing "AB=" should be allowed meaning an empty value?

```
  lines = kvline_parse("AB=", KV_QUOTED);
  tt_assert(lines);
  tt_str_op(lines->key, OP_EQ, "AB");
  tt_str_op(lines->value, OP_EQ, "");

  lines = kvline_parse("AB=", 0);
  tt_assert(lines); <--- FAILURE IS HERE
  tt_str_op(lines->key, OP_EQ, "AB");
  tt_str_op(lines->value, OP_EQ, "");

```

File: src/lib/encoding/kvline.h [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+/**
+ * \file kvline.h
+ *
+ * \brief Header for kvline.c
+ **/
+
+#ifndef TOR_KVLINE_H
+#define TOR_KVLINE_H
+
+struct smartlist_t;
```
Comment (dgoulet-tor) [inline src/lib/encoding/kvline.h]: Don't think this is needed in the header.

File: src/lib/encoding/kvline.c (hunk 4 above)
Comment (nmathewson) [reply src/lib/encoding/kvline.c]: That's right; I've added a clarifying comment

**Pull request closed without merging**